Add tee support for readable byte streams

youennf opened the pull request:
#### df0241e3ab24bb75fd5513ae0fcda4a71fc36be3
<pre>
Add tee support for readable byte streams
<a href="https://rdar.apple.com/161246220">rdar://161246220</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299440">https://bugs.webkit.org/show_bug.cgi?id=299440</a>

Reviewed by Chris Dumez.

We add support for teeing byte streams, following <a href="https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamtee.">https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamtee.</a>
We introduce a class StreamTeeState which stores the various state values needed by this algorithm.
This tee state is kept alive as long as:
- A tee child stream is alive.
- The tee parent stream is not closed.

The tee state keeps a reference to the parent stream and weak references to the child streams.
We currently store as strong values the cancelled reasons until we can cancel the parent stream.

Covered by rebased tests.

Canonical link: <a href="https://commits.webkit.org/300827@main">https://commits.webkit.org/300827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eea065150b15031c8f700dfc2a3d988a376be728

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34393 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130812 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76133 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ff5dc97e-4b1b-4e68-837c-9062a480e955) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125858 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44420 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52290 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/07a7d5a4-eefd-449c-9ded-602a4380596c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126935 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110906 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6b9fb3dc-35c8-4cdb-af1d-3ec879a68fb7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29069 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74283 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105126 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133478 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50934 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38797 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51308 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107126 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26100 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47944 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26194 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47794 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50787 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56551 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50261 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53607 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51935 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/763047d11d2856241ea713ca5efd6fa02934db3d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51569

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/76177e44f1491ef3316e7d7af90d4b377397691a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51569

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f43138c84922e7eefc7b3972e9e5f04a1977a5c0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51569

cdumez commented on Source/WebCore/Modules/streams/StreamTeeUtilities.cpp:
```diff
@@ -0,0 +1,556 @@
+    RefPtr<ReadableStream> m_branch1;
+    RefPtr<ReadableStream> m_branch2;
+
+    // FIXME: we should probably have m_stream mark m_branch1Reason and m_branch2Reason instead of taking strong.
```
Indeed.

youennf committed:
Add tee support for readable byte streams
rdar://161246220
https://bugs.webkit.org/show_bug.cgi?id=299440

Reviewed by Chris Dumez.

We add support for teeing byte streams, following https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamtee.
We introduce a class StreamTeeState which stores the various state values needed by this algorithm.
This tee state is kept alive as long as:
- A tee child stream is alive.
- The tee parent stream is not closed.

The tee state keeps a reference to the parent stream and weak references to the child streams.
We currently store as strong values the cancelled reasons until we can cancel the parent stream.

Covered by rebased tests.

Canonical link: https://commits.webkit.org/300827@main

webkit-commit-queue commented:
Committed 300827@main (df0241e3ab24): <https://commits.webkit.org/300827@main>

Reviewed commits have been landed. Closing PR #51569 and removing active labels.
